interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp

```diff
@@ -0,0 +1,302 @@
+
+    DeclContext *parentDeclContext = IDeclContext->second;
+
+    ASTContext &fromASTContext =
+          Decl::castFromDeclContext(parentDeclContext)->getASTContext();
+    ASTContext &toASTContext =
+          Decl::castFromDeclContext(childCurrentDeclContext)->getASTContext();
+
+    DeclContext::lookup_result lookup_result =
+                                    parentDeclContext->lookup(parentDeclName);
+
+    // Check if we found this Name in the parent interpreter
+    if (!lookup_result.empty()) {
+      // Do the import
+      if (Import(lookup_result, fromASTContext, toASTContext,
+                 childCurrentDeclContext, childDeclName, parentDeclName))
+        return true;
+    }
+
+    return false;
+  }
+
+  ///\brief Make available to child all decls in parent's decl context
+  /// that corresponds to child decl context.
+  void ExternalInterpreterSource::completeVisibleDeclsMap(
+                                const clang::DeclContext *childDeclContext) {
+    assert (childDeclContext && "No child decl context!");
+
+    if (!childDeclContext->hasExternalVisibleStorage())
+      return;
+
+    // Search in the map of the stored Decl Contexts for this
+    // Decl Context.
+    std::map<const clang::DeclContext *, clang::DeclContext *>::iterator
+                  IDeclContext = m_ImportedDeclContexts.find(childDeclContext);
+    // If childCurrentDeclContext was found before and is already in the map,
+    // then do the lookup using the stored pointer.
+    if (IDeclContext == m_ImportedDeclContexts.end()) return ;
+
+    DeclContext *parentDeclContext = IDeclContext->second;
+
+    ASTContext &fromASTContext =
+          Decl::castFromDeclContext(parentDeclContext)->getASTContext();
+    ASTContext &toASTContext =
+          Decl::castFromDeclContext(childDeclContext)->getASTContext();
+
+    // Cling's ASTImporter
+    ClingASTImporter importer = CreateClingASTImporter(toASTContext,
+                                                       fromASTContext,
+                                                       m_ChildInterpreter,
+                                                       m_ParentInterpreter,
+                                                       *this);
+
+    // Filter the decls from the external source using the stem information
+    // stored in Sema.
+    StringRef filter =
+      m_ChildInterpreter->getCI()->getPreprocessor().getCodeCompletionFilter();
+    for (DeclContext::decl_iterator IDeclContext =
+                                      parentDeclContext->decls_begin(),
+                                    EDeclContext =
+                                      parentDeclContext->decls_end();
+                              IDeclContext != EDeclContext; ++IDeclContext) {
+      if (NamedDecl* parentDecl = llvm::dyn_cast<NamedDecl>(*IDeclContext)) {
+        DeclarationName childDeclName = parentDecl->getDeclName();
+        if (childDeclName.getAsIdentifierInfo()) {
```

Comment:
Can you do `if (auto II = childDeclName.getAsIdentifierInfo())` and then use `II` in the next line?